From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Travis Holloway <t.holloway@cpanel.net>
Date: Thu, 23 Sep 2021 15:12:56 +0000
Subject: [PATCH 21/21] Correctly decrement active_daemons also in the case
 that the child process decides on its own to die because of
 MaxRequestsPerChild.

---
 server/mpm/event/event.c | 5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git a/server/mpm/event/event.c b/server/mpm/event/event.c
index 57d7530..6503e0e 100644
--- a/server/mpm/event/event.c
+++ b/server/mpm/event/event.c
@@ -2809,6 +2809,10 @@ static void perform_idle_server_maintenance(int child_bucket, int num_buckets)
         }
         ps = &ap_scoreboard_image->parent[i];
         if (ps->pid != 0) {
+                if (ps->quiescing == 1) {
+                    ps->quiescing = 2;
+                    active_daemons--;
+                }
             for (j = 0; j < threads_per_child; j++) {
                 ws = &ap_scoreboard_image->servers[i][j];
                 status = ws->status;
@@ -2888,7 +2892,6 @@ static void perform_idle_server_maintenance(int child_bucket, int num_buckets)
             ap_mpm_podx_signal(all_buckets[child_bucket].pod,
                                AP_MPM_PODX_GRACEFUL);
             retained->idle_spawn_rate[child_bucket] = 1;
-            active_daemons--;
         } else {
             ap_log_error(APLOG_MARK, APLOG_TRACE5, 0, ap_server_conf,
                          "Not shutting down child: total daemons %d / "
